docs/apispec-docs/src/main/scala/sttp/tapir/docs/apispec/schema/JsonSchemas.scala

```diff
@@ -0,0 +1,44 @@
+package sttp.tapir.docs.apispec.schema
+
+import sttp.apispec.{Schema => ASchema}
+import sttp.tapir.internal.IterableToListMap
+import sttp.tapir.{Schema => TSchema}
+import scala.collection.immutable.ListMap
+
+object JsonSchemas {
+
+  private val toKeyedSchemas = new ToKeyedSchemas
+
+  def apply(
+      schema: TSchema[_],
+      markOptionsAsNullable: Boolean,
+      metaSchema: MetaSchema = MetaSchemaDraft04,
+      schemaName: TSchema.SName => String = defaultSchemaName
+  ): ASchema = {
+
+    val asKeyedSchemas = toKeyedSchemas(schema).drop(1)
+    val keyedSchemas = ToKeyedSchemas.unique(asKeyedSchemas)
+
+    val keysToIds = calculateUniqueIds(keyedSchemas.map(_._1), (key: SchemaKey) => schemaName(key.name))
+    val toSchemaReference = new ToSchemaReference(keysToIds, refRoot = "#/$defs/")
+    val tschemaToASchema = new TSchemaToASchema(toSchemaReference, markOptionsAsNullable)
+    val keysToSchemas = keyedSchemas.map(td => (td._1, tschemaToASchema(td._2))).toListMap
+    val schemaIds = keysToSchemas.map { case (k, v) => k -> ((keysToIds(k), v)) }
+
+    val nestedKeyedSchemas = (schemaIds.values)
+    // TODO proper handling of ref input schema
```

Comment:
**thought:** Should we just make the method return a `ReferenceOr[ASchema]` and let the user handle this?